File: src/solutions/day16/day16.py
import heapq

def find_index(maze, t):
    for i, r in enumerate(maze):
        for j, c in enumerate(r):
            if c == t:
                return (i, j)

with open('sample.txt') as f:
    maze = [l for l in f.readlines()]

rows, cols = len(maze), len(maze[0])
r, c = find_index(maze, "S")

start_r, start_c = find_index(maze, "S")
result = float("inf")

pq = [(0, start_r, start_c, 0 , 1)]
seen = {(start_r, start_c, 0, 1)}

while pq:
    s, r, c, dr, dc = heapq.heappop(pq)
    seen.add((r, c, dr, dc))

    if maze[r][c] == "E":
        print("Solution: ", s)
        break

    for new_s, new_r, new_c, new_dr, new_dc in [(s + 1, r + dr, c + dc, dr, dc), (s + 1000, r, c, dc, -dr), (s + 1000, r, c, -dc, dr)]:
        if 0 <= new_r < rows and 0 <= new_c < cols:
            if maze[new_r][new_c] == "#": continue
            if (new_r, new_c, new_dr, new_dc) in seen: continue
            heapq.heappush(pq, (new_s, new_r, new_c, new_dr, new_dc))


    
    

File: src/solutions/day16/day16.ts
import { readFileSync } from 'node:fs';
import Heap from 'heap-js';

const findStart = (maze) => {
    for (let r = 0; r < maze.length; r++) {
        for (let c = 0; c < maze[r].length; c++) {
            if (maze[r][c] === "S") {
                return [r, c];
            }
        }
    }
    return null;
};

const maze_str = readFileSync('sample.txt', 'utf-8');
const rows = maze_str.split('\n').filter(row => row.trim() !== '');
const maze = rows.map(r => r.trim().split(""));

const start = findStart(maze);
if (!start) {
    console.error("Error: Start point 'S' not found in the maze.");
    process.exit(1);
}
const [r, c] = start;
const pq = new Heap((a, b) => a[0] - b[0]);
pq.push([0, r, c, 0, 1, [[r, c]]]);

const seen = new Set();
seen.add(`${r},${c},0,1`);

const seats = new Set();

while (pq.size() > 0) {
    let [s, r, c, dr, dc] = pq.pop();

    if (maze[r][c] === "E") {
        console.log("Solution ", s);
        break;
    }

    for (let [newScore, newR, newC, newDr, newDc] of [
        [s + 1, r + dr, c + dc, dr, dc],
        [s + 1000, r, c, dc, -dr],
        [s + 1000, r, c, -dc, dr]
    ]) {
        if (0 <= newR && newR < maze.length && 0 <= newC && newC < maze[0].length) {
            if (maze[newR][newC] === "#") continue;

            const stateKey = `${newR},${newC},${newDr},${newDc}`;
            if (seen.has(stateKey)) continue;

            pq.push([newScore, newR, newC, newDr, newDc]);
            seen.add(stateKey);
        }
    }
}
